core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java

```diff
@@ -481,6 +481,18 @@ public TrimResult trimFields(
     final int fieldCount = rowType.getFieldCount();
     final RelNode input = project.getInput();
 
+    boolean containsSubQuery = false;
+    for (RexNode node : project.getProjects()) {
+      if (RexUtil.containsSubQuery(node)) {
+        containsSubQuery = true;
+        break;
+      }
+    }
+    // Do not trim Project's fields before SubQueryRemoveRule applies.
+    if (containsSubQuery) {
+      return result(project, Mappings.createIdentity(fieldCount));
+    }
+
```

Comment:
Does this code has test coverage? I don't see any plan changes due to this. Also the effects of this change are much broader than the changes in `RelBuilder` thus I would suggest to treat this separately in another JIRA.

Reply:
The test added in `sub-query.iq` will fail without this change. It surprised me too that this does not affect any existing plan tests. I guess that's because we don't have enough plan tests for sub query. If there is a test case which is affected by this change, then it was already wrong.

Due to the integrity of this issue I think this is ok to keep it in the same PR, WDYT?

Reply:
I didn't notice that the new test in `sub-query.iq` was passing from here, great! Since there are tests then I guess it is OK to keep the changes as part of this PR.

> If there is a test case which is affected by this change, then it was already wrong.
It may be wrong but not always, right? This is a rough check as we discussed not precise.

Moreover if there is a sub-query in the top project then field trimmer will stop without doing anything. I think we can do better than that. Is it valid to trim its input? I mean something like the following:

```java
if (containsSubQuery) {
  return trimFields((RelNode) project, fieldsUsed, extraFields);
}
```

Reply:
Maybe it is worth adding the new test also in `RelFieldTrimmerTest` since it is explicitly targeting this code path.

Reply:
tests added.